Amortised validator reward payouts

mzxyz opened the pull request:
Instead of triggering reward payout / balance transfer at every extrinsic, we want to store the rewards and make a payout once per era (per validator) for performance reasons.

## Implementation

Add:
- `CurrentEraTransactionRewards` storage in staking module
- add, set and get (behind test) methods for the storage above
- `fn split_rewards_evenly_to_all(recipients, total_amount)` to split total_amount (rewards) evenly across the recipients
- call the above method inside `fn new_era` responsible for payout at era change

Change:
- refactor `fn on_nonzero_unbalanced` from runtime/src/impls.rs so that transaction fee rewards are accumulated in the `CurrentEraTransactionRewards` storage

## Test
Config updates:
- Add `start_session` and `start_era` functions to trigger session and era manually.

Test cases:
- new session and new era work
-  transaction rewards storage update works
- validators received staking reward after new era start
- validators should received transaction reward after new era start

mzxyz committed:
refactor on_nonzero_unbalanced for SplitToAllValidators
[ci skip]

mzxyz committed:
add new imlementation of tx reward
[ci skip]

mzxyz committed:
try to trigger session and era in test
[ci skip]

roy-powerfinance committed:
Added logic to adjust total issuance of the reward currency after the rewards
are paid out to the validators at the start of a new era
Also ran cargo fmt

jordy25519 commented on crml/staking/src/lib.rs:
```diff
@@ -726,6 +726,9 @@ decl_storage! {
 		CurrentEraPointsEarned get(fn current_era_reward): EraPoints;
 
+		/// Total transaction payment rewards for elected validators
+		pub CurrentEraTransactionRewards get(fn current_era_transaction_rewards): BalanceOf<T>;
```
Can we create a public setter method for this instead? so that other modules cannot access it directly ❓

roy-powerfinance replied:
Added a Getter and an adder, made the storage private

jordy25519 commented on crml/staking/src/lib.rs:
```diff
@@ -1370,6 +1373,32 @@ impl<T: Trait> Module<T> {
 	}
 
+	/// Payout transaction rewards to all the validators. Called at the beginning of an era
+	fn transaction_reward_validators() {
```
Passing in the `validators` and `total_reward` as parameters would make this function nicer

roy-powerfinance replied:
yep did that. Repurposed the function to just paying out some reward evenly to a bunch of people

jordy25519 commented on crml/staking/src/lib.rs:
```diff
@@ -1370,6 +1373,32 @@ impl<T: Trait> Module<T> {
+		sp_std::mem::forget(imbalance);
+
+		// reset chain storage for the total reward for the new era
+		<CurrentEraTransactionRewards<T>>::kill();
```
Is it possible to use a `<CurrentEraTransactionRewards<T>>::take()` instead?

hoani replied:
^ This is probably a better way to go than my above comment on using `::get()`

roy-powerfinance replied:
Used Take instead of get and kill

jordy25519 reviewed (commented):
Looks really good.
Some general tidy-ups suggested

hoani commented on crml/staking/src/lib.rs:
```diff
@@ -1370,6 +1373,32 @@ impl<T: Trait> Module<T> {
+	/// Payout transaction rewards to all the validators. Called at the beginning of an era
+	fn transaction_reward_validators() {
+		let validators = Self::current_elected();
+		let total_reward = Self::current_era_transaction_rewards();
```
If we want to avoid making the getter for this public (which I think is what Jordy is saying above) we can use `< CurrentEraTransactionRewards <T>>::get()` instead of `Self::current_era_transaction_rewards()`

roy-powerfinance replied:
Removed the getter function.

mzxyz committed:
refactor runtime test mock
[ci skip]

cowboy-bebug commented on runtime/tests/mock.rs:
```diff
@@ -93,14 +322,14 @@ impl ExtBuilder {
 
 		let mut t = frame_system::GenesisConfig::default()
-			.build_storage::<Runtime>()
+			.build_storage::<Test>()
```
Something to note is: by passing in Test instead of Runtime, we are no longer testing the actual Runtime struct <- which was the original purpose I think

cowboy-bebug replied:
Had a discussion offline. We will keep using `Test` until we can correctly trigger session end based on Babe

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1370,6 +1373,32 @@ impl<T: Trait> Module<T> {
+
+		let reward = total_reward / validators_len;
+		for v in validators.iter() {
+			Self::make_payout(&v, reward);
```
Do we need to handle PositiveImbalance returned from make_payout?

mzxyz committed:
fix mock config error
[ci skip]

roy-powerfinance committed:
Made some improvement according to comments in the PR

roy-powerfinance committed:
fixed some build errors

mzxyz committed:
fmt codebase

mzxyz committed:
try to fix test
[ci skip]

cowboy-bebug committed:
Add a get method for Reward storage and related test

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1373,6 +1373,11 @@ impl<T: Trait> Module<T> {
 		imbalance
 	}
 
+	#[cfg(any(feature = "std", test))]
```
Not sure if this is frowned upon to have a function guarded behind test. But there are some examples from plug here:: https://github.com/plugblockchain/plug-blockchain/blob/develop/frame/system/src/lib.rs#L853

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1373,6 +1373,11 @@ impl<T: Trait> Module<T> {
+	#[cfg(any(feature = "std", test))]
+	pub fn get_current_era_transaction_fee_reward() -> BalanceOf<T> {
+		CurrentEraTransactionRewards::<T>::get()
+	}
```
Not sure if this is frowned upon to have a function guarded behind test. But there are some examples from plug here:: https://github.com/plugblockchain/plug-blockchain/blob/develop/frame/system/src/lib.rs#L853

hoani replied:
I agree with  @cowboy-bebug, but cannot find a nice way of exposing this without a public getter.
This will have to do unless we find a better solution

mzxyz committed:
tx payment test passed

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -136,6 +136,7 @@ fn start_session(session_index: SessionIndex) {
 	for i in Session::current_index()..up_to_session_index {
 		System::set_block_number((i + 1).into());
 		pallet_babe::CurrentSlot::put(Babe::current_slot() + EpochDuration::get());
+		Timestamp::set_timestamp((System::block_number() * 1000).into());
```
👍 what do u think about using `ExpectedBlockTime::get()` (which is configured as 3 seconds) instead of 1000?

mzxyz replied:
This change should be good, but when do the replacement,  all the test for `rewards` failed. Not sure what's the problem

cowboy-bebug replied:
Yeah they become off by just 1 and I'm not sure why that happens yet.

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -151,16 +152,34 @@ fn start_era(era_index: EraIndex) {
 	crml_staking::inflation::compute_total_payout(
 		<Runtime as crml_staking::Trait>::RewardCurve::get(),
-		<Staking>::slot_stake() * 2,
+		<crml_staking::SlotStake<Runtime>>::get() * validator_count,
```
`SlotStake` has a public get method so this also works:
```rust
Staking::slot_stake() * validator_count,
```

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -151,16 +152,34 @@ fn start_era(era_index: EraIndex) {
 }
 
-pub fn current_total_payout_for_duration(duration: u64) -> Balance {
+fn current_total_payout(validator_count: Balance) -> Balance {
```
👍

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -151,16 +152,34 @@ fn start_era(era_index: EraIndex) {
+			.iter()
+			.map(|v| (v.0.clone(), 1))
+			.collect::<Vec<(AccountId, u32)>>(),
+	);
```
Can we leave a comment to explain what this `reward_by_ids` is doing?
[Optional] can refactor into something like:
```rust
// Mapping of stash account and points to be assigned to them
let validators_points: Vec<(AccountId, u32)> = validators
    .iter()
    .map(|v| (v.0.clone(), 1))
    .collect();
Staking::reward_by_ids(validators_points); // CurrentEraPointsEarned storage is updated with new points
```

cowboy-bebug committed:
Fix a bug where CurrentEraPointsEarned is cleared every session
Move CurrentEraPointsEarned::take() to where it’s consumed
(else it’s going to be destroyed on every session change)
Remove a redundant line
Add pre-payout balance check for tests
Improve transaction reward storage test

cowboy-bebug commented on crml/staking/src/lib.rs:
```diff
@@ -1445,7 +1443,6 @@ impl<T: Trait> Module<T> {
 			let total_rewarded_stake = Self::slot_stake() * validator_len;
 
 			// Pay the accumulated tx fee as rewards to all validators
-			let validators = Self::current_elected();
```
Removed because this was a duplicated line with Line 1441

mzxyz committed:
update codes with comments

hoani commented on crml/staking/src/lib.rs:
```diff
@@ -726,6 +726,9 @@ decl_storage! {
 		CurrentEraPointsEarned get(fn current_era_reward): EraPoints;
 
+		/// Total transaction payment rewards for elected validators
+		CurrentEraTransactionRewards : BalanceOf<T>;
```
[Suggestion] To clearly distinguish between fee rewards and staking rewards, I suggest naming:
`CurrentEraTransactionRewards -> CurrentEraFeeRewards`
and
`split_rewards_evenly_to_all -> split_fee_rewards_evenly_to_all`

mzxyz replied:
Good suggestion, will update soon

hoani commented on runtime/src/impls.rs:
```diff
@@ -49,33 +46,11 @@ pub struct SplitToAllValidators;
-			let _ = total_imbalance.subsume(payout);
+		let amount = imbalance.peek();
+
+		if !amount.is_zero() {
```
Is this check neccessary?
I don't think we need to worry if the amount is zero

mzxyz replied:
@royyang-centrality actually for me, I'm not sure about this, if we remove the check, there is no problem, it is just prevent some unnecessary state calculation

hoani commented on runtime/tests/tests.rs:
```diff
@@ -146,11 +147,40 @@ fn advance_session() {
+	let previous_era_start = <crml_staking::CurrentEraStart<Runtime>>::get();
+	let era_duration = now - previous_era_start;
+
+	crml_staking::inflation::compute_total_payout(
```
to make this more readable, can we do:
`let (payout, _) = crml_staking::inflation::compute_total_...`
and then return `payout` instead of `.0`?

hoani commented on runtime/tests/tests.rs:
```diff
@@ -183,11 +213,64 @@ fn start_era_works() {
+
+			initialize_block();
+			start_era(1);
+			advance_session(); // advance a session to trigger the beginning of era 2
```
Isn't this starting era 1, not era 2? (referring to comments)

mzxyz replied:
The comment should be changed to era 1

cowboy-bebug replied:
The comment is correct: `assert_eq!(Staking::current_era(), 2);` passes after `advance_session()`. I'll add this assert and revert the comment back to the original

hoani commented on runtime/tests/tests.rs:
```diff
@@ -183,11 +213,64 @@ fn start_era_works() {
+			// At the start of the next era (13th session), transaction rewards should be cleared (and paid out)
+			start_era(2);
+			advance_session();
+			assert_eq!(Staking::get_current_era_transaction_fee_reward(), 0);
```
Would be good to check validator balances here, if possible?
From https://github.com/cennznet/cennznet/blob/ada8ea4eb35cc861c8e0a1185b7d3f66837b6bb2/runtime/tests/mock.rs#L32
it looks like the default validators are:
"Alice", "Bob", "Charlie" (it takes 3 by default)

hoani replied:
If we do check these accounts, we might need to make Dave, Eve and Ferdie do the transactions to prevent Alice, Bob and Charlie from having their accounts changed by fees.

mzxyz replied:
In my opinion, there has `balance check` before and after 2 era in `staking_genesis_config_works`, for this test is just check whether `transaction_rewards_storage` working properly.

hoani replied:
Discussed with @HaoZhaiX .
This is addressed in `staking_validators_should_receive_equal_transaction_fee_reward `

hoani commented on runtime/tests/tests.rs:
```diff
@@ -261,17 +387,21 @@ fn staking_validators_should_receive_equal_transaction_fee_reward() {
 			for validator in validators {
 				// Check tx fee reward went to the stash account of validator
 				assert_eq!(
 					<GenericAsset as MultiCurrencyAccounting>::free_balance(&validator.0, Some(CENTRAPAY_ASSET_ID)),
```
Can we use the same pattern:
`let (stash, _) = validator;`
as above to keep the checks consistent?

hoani commented on runtime/tests/tests.rs:
```diff
@@ -261,17 +387,21 @@ fn staking_validators_should_receive_equal_transaction_fee_reward() {
 				assert_eq!(
 					<GenericAsset as MultiCurrencyAccounting>::free_balance(&validator.0, Some(CENTRAPAY_ASSET_ID)),
-					balance_amount + fee_reward
+					balance_amount + per_fee_reward + per_staking_reward
```
Can we use the same pattern: `let (stash, _) = validator;` as above to keep the checks consistent?

hoani reviewed (changes requested):
Overall looks good.
I would like a couple of checks to stash account balances added to the `current_era_transaction_rewards_storage_update_works` test.
Also a few nitpicks and suggestions - up to you if you want to follow them

hoani reviewed (approved):
Discussed changes with @HaoZhaiX, changes look good to me, the remainder of my comments are suggestions or nitpicks.

mzxyz committed:
update codes with last comments and refactor free balance assert

roy-powerfinance commented on runtime/tests/tests.rs:
```diff
@@ -242,37 +366,31 @@ fn staking_validators_should_receive_equal_transaction_fee_reward() {
-			let remainder = fee % validators.len() as u128;
+			let per_fee_reward = fee / validators.len() as Balance;
 
-			let previous_total_issuance = GenericAsset::total_issuance(&CENTRAPAY_ASSET_ID);
```
Optional:
I think we should check if the total issuance is adjusted correctly in this test
I.e. total issuance in the new era = previous_issuance + total_staking_reward  - transaction_fee_remainder - staking_reward_remainder

cowboy-bebug replied:
Thanks for pointing this out. This will be addressed

roy-powerfinance reviewed (changes requested):
Discussed with Eric - I suggested that we add a total issuance check to the test just to be safe.
Otherwise this PR LGTM

cowboy-bebug committed:
Add staking_reward_remainder_should_work, which could be removed after resolving an issue
Add alias MultiCurrencyAccounting -> MultiCurrency to save key strokes
Add more asserts
Add TODOs for total_issuance checks

cowboy-bebug committed:
Add partially working total issuance check

cowboy-bebug committed:
Add RewardFees event for fee payout so we can keep track later
Refactor BalanceOf<T> typing
Allow array_into_iter to remove warning
Run formatter

cowboy-bebug commented on runtime/tests/tests.rs:
```diff
@@ -419,23 +325,27 @@ fn staking_reward_should_work() {
+			assert_eq!(GenericAsset::total_issuance(CENNZ_ASSET_ID), total_issuance);
+			assert_eq!(
+				GenericAsset::total_issuance(CENTRAPAY_ASSET_ID), // FIXME: 33_000_000 is coming from a weird timing issue between sessions
+				total_issuance + inflation + 33_000_000, // FIXME: Changing era_duration in fn current_total_payout outputs different inflation amount (but not sure what it should be)
```
Tried to figure out why we end up issuing `33_000_000` much more cennz (which is more than the max_payout), but still not sure. It's definitely something to do with the `era_duration` in `fn current_total_payout` (changing this results in different inflation amount).

roy-powerfinance replied:
If you refer to the fn new_era(), the total_reward (first parameter returned from compute_total_payout()) is what is actually used to reward the validators. The max is only used to ensure the total_payout doesn't go over the max. 
therefore I think the test here should use total_reward to calculate total issuance 
new_issuance = total_reward / validator_len * validator_len; //this also takes care of remainders

roy-powerfinance replied:
what's leftover from given out is actually burned
```let rest = _max_payout.saturating_sub(total_payout);
Self::deposit_event(RawEvent::Reward(total_payout, rest));```

cowboy-bebug replied:
What's leftover is *not* burned. It's issued using T::RewardCurrency (cpay) back to the T::RewardRemainder (which is currently configured as treasury):
https://github.com/cennznet/cennznet/blob/19ca3f16c6ceb9f56df388d320bfc002a32c50cc/crml/staking/src/lib.rs#L1439
(also see this commit that shows the related system events: https://github.com/cennznet/cennznet/pull/153/commits/21791f30e81573fb46af086e7d6bd8e272ac6f37)

------

https://github.com/cennznet/cennznet/blob/19ca3f16c6ceb9f56df388d320bfc002a32c50cc/crml/staking/src/lib.rs#L1413
^ The variable names are confusing but `max_payout` not only ensures that you don't pay over that amount, but also it's the inflation target that we are trying to achieve. Hence `total_payout` is paid out to the validators and nominators and then the `rest` is calculated (max_payout - total_payout) and issued to the `T::RewardRemainder`:
https://github.com/cennznet/cennznet/blob/19ca3f16c6ceb9f56df388d320bfc002a32c50cc/crml/staking/src/lib.rs#L1435-L1439

cowboy-bebug replied:
And if it helps the numbers are:
```rust
let (total_reward, inflation) = reward_validators(&validators);
assert_eq!(total_reward, 711_000_000);
assert_eq!(inflation,  1_896_000_000);
```

hoani replied:
I'm gonna look into this a bit deeper.
@cowboy-bebug makes some interesting points - it might be that the currency overall should have a fixed inflation, but we will reward less than that

roy-powerfinance replied:
Uhh yes you are right! the unbalance is implemented for the treasury:
impl<T: Trait> OnUnbalanced<NegativeImbalanceOf<T>> for Module<T>

and it puts the money into the treasury account:
let _ = T::Currency::resolve_creating(&Self::account_id(), amount);

I'm surprised at the difference between the two numbers though - like you said, it may have something to do with the duration of the era.

cowboy-bebug replied:
I got a nice tip from @hoani to try out and resolve this issue. I'll mark it as resolved for now 😄

roy-powerfinance replied:
As discussed in team planning, fixing this is split into another task, and will be done in another PR.

cowboy-bebug committed:
Resolve rebase conflicts
